### DISCUSSION TITLE
Keybind for cycling through layout engines

### DISCUSSION THREAD

**Discussion by sun-wise-man** [dalyIsaac/Whim#797] [Q&A] [answered]
Is there a way to set up keybind for cycling layout engine?

**Comment by dalyIsaac** [comment, marked as answer]
I've just made a new release with two commands built in, as I'm planning to update how this works a bit in a future release.

With the new built-in commands, you can do:

```csharp
context.KeybindManager.SetKeybind("whim.core.cycle_layout_engine.next", new Keybind(KeyModifiers.LControl | KeyModifiers.LShift | KeyModifiers.LAlt, VIRTUAL_KEY.VK_L));
context.KeybindManager.SetKeybind("whim.core.cycle_layout_engine.previous", new Keybind(KeyModifiers.LControl | KeyModifiers.LShift | KeyModifiers.LAlt | KeyModifiers.LWin, VIRTUAL_KEY.VK_L));
```

If you were planning to create custom commands, you'd do something like this (for now, until #796 is merged):

```csharp
context.CommandManager.Add("activate_next_layout", "Activate next layout", () => context.WorkspaceManager.ActiveWorkspace.CycleLayoutEngine());
context.CommandManager.Add("activate_previous_layout", "Activate previous layout", () => context.WorkspaceManager.ActiveWorkspace.CycleLayoutEngine());

context.KeybindManager.SetKeybind("whim.custom.activate_next_layout", new Keybind(KeyModifiers.LControl | KeyModifiers.LShift | KeyModifiers.LAlt, VIRTUAL_KEY.VK_L));
context.KeybindManager.SetKeybind("whim.custom.activate_previous_layout", new Keybind(KeyModifiers.LControl | KeyModifiers.LShift | KeyModifiers.LAlt, VIRTUAL_KEY.VK_H));
```

**Reply by sun-wise-man** [reply]
Ok, got it! Thank you. I just added it to my config, and it works.

A little bit of a follow-up question, with a case like this:
I have two windows opened in the tree layout: Firefox and Notepad. With a focused window on Firefox. When I change the layout to focus, Firefox is not the focused window in the focus layout. Instead, it opens the notepad in the focus layout.

Is there a functionality to bring a focused window in another layout to open as a focused window in the focus layout? In that case, Firefox, because that's my focused window in tree layout.

**Reply by dalyIsaac** [reply]
This functionality doesn't currently exist in an any way I'd be happy recommending. It's something I'd like to improve - it's tracked in #785.
